### PR TITLE
Add melange tests

### PR DESCRIPTION
**Pull request by jchavarri**
Adds a few tests for melange mode. All of them without using `public_name` in libraries as installation is not supported yet.

- [x] simple library
- [x] mli files
- [x] reason syntax
- [x] with multiple files (to detect potential race conditions due to untracked cmi deps)
~- [ ] private modules~ Edit: for now left out of scope, as they will be tackled together with library installation.

### CODE REVIEW

File: test/blackbox-tests/test-cases/melange/multilib.t/lib/m_1_1_1_2.ml [hunk 1]
```diff
@@ -0,0 +1,3 @@
...XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX *)
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/multilib.t/lib/m_1_1_1_2.ml:1]: do we really need this long comment?

File: test/blackbox-tests/test-cases/melange/multilib.t/lib/m_1_1_1_3.ml [hunk 2]
```diff
@@ -0,0 +1,3 @@
...XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX *)
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/multilib.t/lib/m_1_1_1_3.ml:1]: ditto

File: test/blackbox-tests/test-cases/melange/multilib.t/run.t [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Compilation using melange with a library
+The library in the test includes tens of files to make sure cmjs are read after they are written.
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/multilib.t/run.t:2]: Why do we need so many files for this test? If it's some sort of benchmark, then it shouldn't be part of the test suite.

File: test/blackbox-tests/test-cases/dune [hunk 4]
```diff
@@ -28,6 +28,11 @@
   (deps %{bin:node})
   (alias runtest-js)))
 
+(subdir
+ melange
+ (cram
+  (alias runtest-melange)))
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/dune:34]: the tests that execute JS should also depend on node.

File: test/blackbox-tests/test-cases/melange/reason.t/lib/x.re [hunk 5]
```diff
@@ -0,0 +1,3 @@
+let buy_it = "buy " ++ Y.it
+
+let () = print_endline(buy_it)
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/reason.t/lib/x.re:3]: We don't actually depend on reason for our tests (too unreliable of a dependency) and make instead use a fake refmt binary. We should use that binary for these tests as well.

File: test/blackbox-tests/test-cases/melange/reason.t/dune-project [hunk 6]
```diff
@@ -0,0 +1,2 @@
+(lang dune 3.5)
+(using melange 0.1)
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/reason.t/dune-project:2]: if you define the package in the project file, you can skip the empty opam file.

File: test/blackbox-tests/test-cases/melange/multilib.t/lib/m_1_1_1_2.ml (hunk 1 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/multilib.t/lib/m_1_1_1_2.ml]: not really, removed in https://github.com/jchavarri/dune/pull/5/commits/5da395db6ff91afb8025721de69ca2a16dd6d48d.

File: test/blackbox-tests/test-cases/melange/multilib.t/lib/m_1_1_1_3.ml (hunk 2 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/multilib.t/lib/m_1_1_1_3.ml]: https://github.com/jchavarri/dune/pull/5/commits/5da395db6ff91afb8025721de69ca2a16dd6d48d

File: test/blackbox-tests/test-cases/melange/multilib.t/run.t (hunk 3 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/multilib.t/run.t]: This test was added to check some bug where cmjs were not added as deps when one lib depends on another. I originally thought the error would only happen with large builds, but I just realized it happens with single module libraries 🤦  fixed in https://github.com/jchavarri/dune/pull/5/commits/5da395db6ff91afb8025721de69ca2a16dd6d48d.

File: test/blackbox-tests/test-cases/dune (hunk 4 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/dune:35]: fixed in https://github.com/jchavarri/dune/pull/5/commits/0b9b266fd8bb8815936764524976b5e6fee86330.

File: test/blackbox-tests/test-cases/melange/reason.t/dune-project (hunk 6 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/reason.t/dune-project]: https://github.com/jchavarri/dune/pull/5/commits/11bd20d4edd95b93acca0d7ea1a8e0f6e7608df2

File: test/blackbox-tests/test-cases/melange/reason.t/lib/x.re (hunk 5 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/reason.t/lib/x.re]: After further thinking, I don't believe there's a need to test Reason for the `melange` mode, as it should not be affected.

**Pull request closed without merging**